src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunctionSet.cs

```diff
@@ -0,0 +1,362 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Xml.Linq;
+using Microsoft.PowerFx.Core.Texl.Builtins;
+using Microsoft.PowerFx.Core.Utils;
+
+namespace Microsoft.PowerFx.Core.Functions
+{
+    internal class TexlFunctionSet
+    {
+        // Dictionary key: function.Name
+        private Dictionary<string, List<TexlFunction>> _functions;
+
+        // Dictionary key: function.LocaleInvariantName
+        private Dictionary<string, List<TexlFunction>> _functionsInvariant;
+
+        // Dictionary key: function.Namespace
+        private Dictionary<DPath, List<TexlFunction>> _namespaces;
+
+        // List of all function.GetRequiredEnumNames()
+        private List<string> _enums;
+
+        // Count of functions
+        internal int _count;
+
+        internal Dictionary<string, List<TexlFunction>>.KeyCollection FunctionNames => _functions.Keys;
+
+        internal Dictionary<string, List<TexlFunction>>.KeyCollection InvariantFunctionNames => _functionsInvariant.Keys;
+
+        internal Dictionary<DPath, List<TexlFunction>>.KeyCollection Namespaces => _namespaces.Keys;
+
+        internal IEnumerable<TexlFunction> WithName(string name) => WithNameInternal(name);
+
+        private List<TexlFunction> WithNameInternal(string name) => _functions.TryGetValue(name, out List<TexlFunction> result) ? result : new List<TexlFunction>();                
+
+        internal IEnumerable<TexlFunction> WithName(string name, DPath ns) => _functions.TryGetValue(name, out List<TexlFunction> result) ? result.Where(f => f.Namespace == ns) : new List<TexlFunction>();
+
+        internal IEnumerable<TexlFunction> WithInvariantName(string name) => WithInvariantNameInternal(name);
+
+        private List<TexlFunction> WithInvariantNameInternal(string name) => _functionsInvariant.TryGetValue(name, out List<TexlFunction> result) ? result : new List<TexlFunction>();
```

Comment:
Returning an orphaned mutable collection is risky - it means we have to double check every callsite for improper usage. 

Can we just remove WithInvariantNameInternal() and have callers invoke _functionsInvariant.TryGetValue directly instead?

Reply:
We should not expose our internal structures (_functions and others), so we cannot have the callers use TryGetValue, hence this logic.
Now, if you think returning an empty enumerable is risky, I'll update the code to return null ... but this moves the risk to the caller which now needs to check the IEnumerable isn't null. I think this is worse but as you have a different opinion and already mentioned this "problem", I'll update the code accordingly.

Reply:
After looking at consequences, this change is far too heavy and risky

Reply:
The methods that return a List<T> are private.
All other methods are internal and shouldn't be called by anyone.